examples/pagerank.cpp

```diff
@@ -0,0 +1,181 @@
+// Copyright 2016 Husky Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "boost/tokenizer.hpp"
+
+#include "core/engine.hpp"
+#include "io/input/hdfs_line_inputformat.hpp"
+
+class PRVertex {
+public:
+    float pr;
+    typedef int KeyT;
+    int key;
+
+    PRVertex() {}
+    explicit PRVertex(KeyT key) {
+        this->key = key;
+    }
+
+    inline virtual KeyT const & id() const {
+        return key;
+    }
+
+    virtual void key_init(KeyT key) {
+        this->key = key;
+    }
+
+    virtual void add_neighbors(KeyT nbKey) {
+        this->nbs.push_back(nbKey);
+    }
+
+    virtual std::vector<KeyT> & get_neighbors() {
+        return this->nbs;
+    }
+
+    static uint64_t partition(const KeyT & key) {
+        return key;
+    }
+
+    friend husky::BinStream & operator >> (husky::BinStream & stream, PRVertex & v) {
+        stream >> v.pr >> v.nbs >> v.key;
+        return stream;
+    }
+
+    friend husky::BinStream & operator << (husky::BinStream & stream, PRVertex v) {
+        stream << v.pr << v.nbs << v.key;
+        return stream;
+    }
+
+protected:
+    std::vector<KeyT> nbs;
+};
+
+std::function<void(PRVertex &)> exec_pagerank(husky::ObjList<PRVertex> & obj_list, husky::PushCombinedChannel<float, PRVertex, husky::SumCombiner<float>> & ch) {
+    std::function<void(PRVertex &)> exec_lambda = [&](PRVertex & v) {
+        float sum = 0;
+        sum += ch.get(v);
+        v.pr = 0.15 + 0.85 * sum;
+
+        if (v.get_neighbors().size() == 0) return;
+
+        float pr_contrib = v.pr / v.get_neighbors().size();
+        for (auto & nb : v.get_neighbors()) {
+            ch.push(pr_contrib, nb);
+        }
+    };
+    return exec_lambda;
+}
+
+class Foo {
+public:
+    typedef int KeyT;
+    int key;
+
+    explicit Foo(KeyT & key) {
+        this->key = key;
+    }
+
+    const KeyT & id() const {
+        return key;
+    }
+};
```

Comment:
But I think the id() member function is necessary for objList to find the object.